File: algo/leetcode_312.py
''' Burst ballons
'''

def maxCoins(nums):
    if not nums: return 0
    from collections import defaultdict
    tbl = defaultdict(lambda: 0)

    arr = [1] + nums + [1]
    n = len(arr)
    # Iterate over intervals with length k
    for k in range(2, n):
        # Consider an interval sans boundaries
        #   l, a1, a2, ..., r
        #   ^               ^
        #   |<----- k ----->|
        # As a result, removing any elements within the boundary
        # would not affect anything else outside. Thus the last
        # elements to remove is when there is only (l, a, r).
        for i in range(0, n - k):
            j = i + k
            v_max = 0
            # On the last removal inside the interval
            # Here t is assumed to be the last element left
            # in the interval to pop.
            for t in range(i + 1, j):
                curr = arr[i] * arr[t] * arr[j]
                v_curr = curr + tbl[(i, t)] + tbl[(t, j)]
                v_max = max(v_max, v_curr)

            tbl[(i, j)] = v_max

    return tbl[(0, n - 1)]


def TEST(nums, tgt):
    res = maxCoins(nums)
    if res != tgt:
        print('Error', res, 'but expect', tgt)
    else:
        print('Ok')


TEST([3,1,5,8], 167)
TEST([35,16,83,87,84,59,48,41], 1583373)

File: algo/leetcode_312.cxx
/**
 * Burst ballons
 */

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int maxCoins(vector<int> &nums) {
        if (nums.empty()) return 0;
        // Add boundary values 1 to both sides
        nums.insert(nums.begin(), 1); nums.push_back(1);
        int n = nums.size();
        // Boundary condition: for the interval (i, i+1) => 0
        vector<vector<int>> tbl_intv_max_val(n, vector<int>(n, 0));
        for (int intv_size = 3; intv_size <= n; ++intv_size) {
            for (int i = 0; i + intv_size <= n; ++i) {
                int j = i + intv_size - 1;
                int curr_max = 0;
                for (int k = i + 1; k < j; ++k) {
                    int pref = tbl_intv_max_val[i][k];
                    int post = tbl_intv_max_val[k][j];
                    int curr = nums[i] * nums[k] * nums[j];
                    curr_max = max(curr_max, pref + curr + post);
                }
                tbl_intv_max_val[i][j] = curr_max;
            }
        }
        return tbl_intv_max_val[0][n - 1];
    }
};

Solution sol;

void TEST(vector<int> nums, int expected) {
    int res = sol.maxCoins(nums);
    if (res != expected)
        cout << "FAIL: expect " << expected << " but got " << res << endl;
    else
        cout << "PASS" << endl;
}

int main() {
    TEST({3,1,5,8}, 167);
    TEST({35,16,83,87,84,59,48,41}, 1583373);
}
